feat: add transform / filter / write consumer lambda

olizilla opened the pull request:
Queue consumer lambda to transform items to the current index format, check if they exists in destination table, and write those records that are missing.

Each record is a stringified array of up to 500 BlocksIndex objects from the Scanner.

Any failed writes from the DynamoDB BatchWriteCommand are send to the `unprocessedWritesQueue` for debugging and re-driving.

Any other errors and the message is released back to the queue. If that batch fails 3 times, it is send to the `batchDeadLetterQueue`

License: MIT

olizilla committed:
feat: add scanner and sst skeleton

The `Scanner` lambda does a full table scan of the legacy index table, and sends batches of records to an SQS Queue. Ported from https://github.com/alexdebrie/serverless-dynamodb-scanner

Uses the sst Config defaults to get hold of the current stage form inside the lambda.
see: https://docs.sst.dev/config#default-values

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: self invoking lambda and scan partition

- Fixes permissions to allow lambda to reinvoke itself when it's about to run out of execution time
- Let us pass table scan partition segment parameters

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: update package-lock

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: add transform / filter / write consumer lambda

Queue consumer lambda to transform items to the current index format, check if they exists in destination table, and write those records that are missing.

Each record is a stringified array of up to 500 BlocksIndex objects from the Scanner.

Any failed writes from the DynamoDB BatchWriteCommand are send to the `unprocessedWritesQueue` for debugging and re-driving.

Any other errors and the message is released back to the queue. If that batch fails 3 times, it is send to the `batchDeadLetterQueue`

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
chore: add comment

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla commented:
- The `unprocessedWritesQueue` will contain stringified arrays of `WriteRequest` objects, the marshalled Dynamo DB BulkWriteCommand format
- The `batchDeadLetterQueue` will contain stringified arrays of `BlockIndex` objects, the same format as the `batchQueue`, so anything in the could be re-driven through the consumer lambda if there is no obvious, intrinsic reason why that batch failed

alanshaw commented on packages/functions/src/consumer.ts:
```diff
@@ -1,15 +1,175 @@
-    const batch = JSON.parse(r.body)
-    console.log(`Batch of ${batch.length}`)
+    const items: BlocksIndex[] = JSON.parse(r.body)
+    console.log(`Processing "${items.length}" items`)
```
lol quotes

alanshaw commented on packages/functions/src/consumer.ts:
```diff
@@ -1,15 +1,175 @@
+ * We ignore any UnprocessedItems from the BatchGetItem response. It will
+ * be interpreted as having not existed, and so cause a possibly un-needed write.
+ */
+export async function* checkExists(dstTable: string, client: DynamoDBClient, items: BlocksCarsPosition[]) {
```
```suggestion
export async function* filterExists(dstTable: string, client: DynamoDBClient, items: BlocksCarsPosition[]) {
```

alanshaw reviewed (approved):
Looks great!

olizilla committed:
feat: test consumer core

extracts migrator function from consumer handler impl and tests it against mock dynamodb

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: split out sync transform from async writeIfMissing steps

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
Merge remote-tracking branch 'origin/main' into consumer

olizilla committed:
chore: add missing dep

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: test in CI

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla committed:
feat: tweak consumer batching params

and add docs

License: MIT
Signed-off-by: Oli Evans <oli@protocol.ai>

olizilla commented:
Running with the staging blocks table as the src and a new empty table as the dst, i get 2 interesting issues

1. The dst table complains "ProvisionedThroughputExceededException: The level of configured provisioned throughput for the table was exceeded. Consider increasing your provisioning level with the UpdateTable API."
2. The consumer lambda errors but the messages it is operating on are not marked as visible again. My guess is I am waiting for the 15 visibility timeout until those messages are retried.

<img width="1207" alt="Screenshot 2023-09-07 at 17 03 44" src="https://github.com/web3-storage/migrate-block-index-infra/assets/58871/328da04a-5cac-4b9a-95e5-2d5679863d4a">
